crates/mun_codegen_macros/src/lib.rs

```diff
@@ -27,175 +21,189 @@ pub fn as_value_derive(input: TokenStream) -> TokenStream {
+                        let ir_type = <#ty>::get_ir_type(type_context);
+                        type_context.target_data.get_preferred_alignment(&ir_type) as usize
+                    }}
+                });
+
+                quote! {{
+                    let mut total_size = 0;
+
+                    let field_sizes = vec![ #(#field_sizes),* ];
+                    let field_alignments = vec![ #(#field_alignments),* ];
+
+                    let mut field_paddings: Vec<usize> = field_sizes
+                        .iter()
+                        .zip(field_alignments.iter())
+                        .map(|(size, align)| {
+                            let padded_size = padded_size(*align, total_size);
+                            let padding = padded_size - total_size;
+                            total_size = padded_size + size;
+                            padding
+                        })
+                        .collect();
+
+                    // Add padding for the end of the struct
+                    let max_align = field_alignments.iter().max().cloned().unwrap_or(1);
+                    let padded_size = padded_size(max_align, total_size);
+                    field_paddings.push(padded_size - total_size);
+
+                    field_paddings
+                }}
+            };
+
+            let field_padding_values = field_padding_types.clone();
+
+            // Generate a list of where clauses that ensure that we can cast each field to an
+            // `inkwell::types::BasicTypeEnum`
+            let field_types = struct_data.fields.iter().map(|f| {
+                let ty = &f.ty;
+                quote! {
+                    Into::<inkwell::types::BasicTypeEnum<'ink>>::into(<#ty>::get_ir_type(context))
+                }
+            });
+
+            // Generate a list of where clauses that ensure that we can cast each field to an
+            // `inkwell::values::BasicTypeValue`
+            let field_types_values = struct_data.fields.iter().enumerate().map(|(idx, f)| {
+                let name = f.ident.as_ref().map(|i| quote! { #i }).unwrap_or_else(|| quote! { #idx });
+                quote! {
+                    {
+                        let value = crate::value::AsValueInto::<'ink, inkwell::values::BasicValueEnum<'ink>>::as_value_into(&self. #name, context);
+                        // eprintln!("- {:?}", value.get_type());
+                        value
                     }
-                    None => (),
-                };
-
-                let struct_ty = context.context.opaque_struct_type(key);
-                context.struct_types.borrow_mut().insert(key, struct_ty);
-                struct_ty.set_body(&[
-                    #(#field_types),*
-                ], false);
-                struct_ty
-            }
-        }
+                }
+            });
+
+            // Generate Phase
+            (quote! {
+                impl<'ink> crate::value::ConcreteValueType<'ink> for #ident {
```

Comment:
It would help me to add some comments here, e.g.:
```suggestion
                // A struct always has an inkwell StructValue as a result type
                impl<'ink> crate::value::ConcreteValueType<'ink> for #ident {
```